Added the resource compliant round robin algorithm

avflor opened the pull request:
fixes #1148 

This branch includes the resource compliant round robin packing algorithm. Some classes were moved to utils so that both resource compliant RR and first fit decreasing use them.

avflor commented:
@kramasamy This version of the round robin packing algorithm will by default return a null packing if the number of containers specified by the user is not sufficient to satisfy the resource requirements of the components. Optionally the user can configure the algorithm (using topologyConfig) to automatically increase the number of containers if the user-specified number is not sufficient so that a packing plan is created. Is this an acceptable behavior? Let me know if you agree with the defaults.

kramasamy commented:
@avflor - the approach is fine with me. @maosongfu and @billonahill can you please check this PR?

kramasamy commented:
One broader question - will it affect the existing tuning of all the jobs?

avflor commented:
It might affect existing jobs if the number of containers provided as input is not sufficient to satisfy the memory requirements of each instance. The current round robin algorithm will reduce the memory requirements of each instance to fit into a smaller number of containers. This algorithm will fail.  If I make the autoadjustment as default then this algo will not affect current jobs. They will not fail but they may use more containers than previously. This is because now the memory requirements of each instance are guaranteed to be satisfied. However with homogeneous containers you may end up seeing more resources being used than the current round robin. 

I think if you want to guarantee the memory resources per instance, this algorithm is better since it provides the guarantee and can automatically figure out the number of containers. However, if you are fine not providing this guarantee but use fewer homogeneous containers, then you should use the existing algo. In case of YARN, I would recommend the enhanced algo since it can support heterogeneous containers and thus will not waste resources.

This algorithm supports configurable padding that is added on top of the container_ram defined in the topology. Thus your users, might see slightly bigger containers as it is not assumed that they take padding into account when defining the container size. This is because the padding value is dependent on the number of instances packed in a container and thus is applied after packing.

 Let me know if you want me to change the default for auto-adjustment.

kramasamy commented:
@avflor - thanks for the detailed explanation. conservatively, can we code this as a separate algorithm - something like Enhanced Round Robin. I can ask a few topologies to turn on the Enhanced Round Robin and see how it works before we roll out widely. If the existing algorithm is changed and we don't its behavior, it will be difficult to roll back - unless we roll back the release. Thoughts?

avflor commented:
@kramasamy I have already coded it as a separate algorithm under the round robin package so that it does not affect existing jobs. We can try it and see. I was thinking whether it makes sense to have the packing algo being configurable for each topology. E.g a topology where minimizing resources is important can use binpacking, another one round robin etc. What do you think?

kramasamy commented:
@avflor - it is definitely possible to configure each topology with a different packing algorithm. All you have to do is during ```heron submit``` pass the option

--config-property "heron.class.packing.algorithm=<classname>"

it should work. let me know if this does not work.

avflor commented:
@kramasamy Cool I'll try that. I thought packing algo is a global config.

kramasamy commented:
@avflor - you can override any config at the command line using ```--config-property``` as long as you know its implications. The only inconvenience is you have to type it every time you submit in the CLI. With this PR #1119, you could setup an .heronrc file in your home directory and set the config property there and you don't have to do in command line every time.

ajorgensen commented on heron/packing/src/java/com/twitter/heron/packing/binpacking/FirstFitDecreasingPacking.java:
```diff
@@ -341,66 +347,29 @@ protected boolean isValidInstance(long instanceRAM) {
-      return Long.compare(ramRequirement, c.ramRequirement) == 0;
+    if (instanceCpu > maxContainerCpu) {
+      LOG.severe(String.format(
+          "This instance requires containers with at least %d vpu vores. The current max container"
```
Sorry is `vpu vores` a typo that should say `cpu cores`?

ajorgensen commented on heron/packing/src/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPacking.java:
```diff
@@ -0,0 +1,419 @@
+import com.twitter.heron.spi.packing.PackingPlan;
+import com.twitter.heron.spi.utils.TopologyUtils;
+
+
```
Extra new lines.

ajorgensen commented on heron/packing/src/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPacking.java:
```diff
@@ -0,0 +1,419 @@
+
+    if (instanceCpu > maxContainerCpu) {
+      LOG.severe(String.format(
+          "This instance requires containers with at least %d vpu vores. The current max container"
```
Same typo here it looks like

ajorgensen commented on heron/packing/src/java/com/twitter/heron/packing/utils/RamRequirement.java:
```diff
@@ -0,0 +1,65 @@
+ */
+public class RamRequirement implements Comparable<RamRequirement> {
+
+
```
extra new lines

ajorgensen commented on heron/packing/src/java/com/twitter/heron/packing/utils/RamRequirement.java:
```diff
@@ -0,0 +1,65 @@
+    RamRequirement c = (RamRequirement) o;
+
+    // Compare the ramRequirement values and return accordingly
+    return Long.compare(ramRequirement, c.ramRequirement) == 0;
```
So this will say that two ram requirements for different components will be equal correct. IE `new RamRequirement("foo", 10).equals(new RamRequirement("bar", 10))` would return true. The hashCode on the other hand takes the component name into account. So while two of these would be equal they would not hash to the same value.

Just want to make sure this is intentional

billonahill commented on heron/packing/src/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPacking.java:
```diff
@@ -0,0 +1,419 @@
+              = topology.getTopologyConfig().getKvsList();
+          boolean adjustNumberOfContainers
+              = Boolean.parseBoolean(TopologyUtils.getConfigWithDefault(
+              topologyConfig, com.twitter.heron.api.Config.TOPOLOGY_ADJUST_NUMBER_CONTAINERS,
```
Do we need TOPOLOGY_ADJUST_NUMBER_CONTAINERS? We could remove it and just state that the contract of this IPacking instance is that it will adjust.

avflor replied:
@billonahill @kramasamy I can remove that but I thought it might be confusing for users to see a larger number of containers than what they declared. That is why I let them clarify the behavior. Do you think it is better to remove that and always adjust if needed?

billonahill replied:
I think it would be reasonable to remove it and sometimes get more containers. In that way the container setting would also be considered a hint.

billonahill commented on heron/packing/src/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPacking.java:
```diff
@@ -0,0 +1,419 @@
+              topologyConfig, com.twitter.heron.api.Config.TOPOLOGY_ADJUST_NUMBER_CONTAINERS,
+              Boolean.toString(false)));
+          if (!adjustNumberOfContainers) {
+            LOG.severe(String.format("Not enough containers to accommodate the instances"));
```
This should contain more detailed info to help the user understand what happened (e.g., how many were needed, how many were requested, etc).

billonahill commented on heron/packing/src/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPacking.java:
```diff
@@ -0,0 +1,419 @@
+            return null;
+          } else {
+            //Automatically adjust the number of containers
+            LOG.info(String.format("Increasing the number of containers"));
```
This should include more info, like increasing from Z to Y for what reason.

billonahill commented on heron/packing/src/java/com/twitter/heron/packing/utils/RamRequirement.java:
```diff
@@ -0,0 +1,65 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.twitter.heron.packing.utils;
```
Typically we don't put model objects in a util package. Could we put the elsewhere and only have static utility helpers under util? A class representing RamRequirement seems like a core object, not a utility.

avflor replied:
@billonahill The container and RamRequirements classes are only used by the packing algorithm. In that sense I don't see them as core components. In the context of packing I'm using them as helper classes that's why I put them under utils. If "utils" is misleading, then how do you think we should name this package?

billonahill replied:
I think it's fine to put these classes directly under c.t.heron.packing if there common objects leveraged by different packing algos.

billonahill commented on heron/packing/src/java/com/twitter/heron/packing/binpacking/FirstFitDecreasingPacking.java:
```diff
@@ -114,10 +117,22 @@ public static String getComponentName(String instanceId) {
+    List<TopologyAPI.Config.KeyValue> topologyConfig = topology.getTopologyConfig().getKvsList();
+    this.maxContainerRam = Long.parseLong(TopologyUtils.getConfigWithDefault(
+        topologyConfig, com.twitter.heron.api.Config.TOPOLOGY_CONTAINER_MAX_RAM_HINT,
+        Long.toString(instanceRamDefault * 4)));
```
4 here and below appears to be a magic number. What does it represent and should it be a constant or a config setting with a default?

billonahill commented on heron/packing/src/java/com/twitter/heron/packing/binpacking/FirstFitDecreasingPacking.java:
```diff
@@ -332,7 +338,7 @@ private int allocateNewContainer(ArrayList<Container> containers) {
    * @return true if the instance is valid, false otherwise
    */
-  protected boolean isValidInstance(long instanceRAM) {
+  protected boolean isValidInstance(long instanceRAM, double instanceCpu, long instanceDisk) {
```
PackingPlan has an inner class Resource, which is a tuple of these 3 objects. Should we make that an outer class and leverage it here?

billonahill commented on heron/packing/src/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPacking.java:
```diff
@@ -0,0 +1,419 @@
+
+  @Override
+  public void close() {
+
```
remove empty line

billonahill commented on heron/packing/tests/java/com/twitter/heron/packing/binpacking/FirstFitDecreasingPackingTest.java:
```diff
@@ -90,6 +89,31 @@ protected PackingPlan getFirstFitDecreasingPackingPlan(TopologyAPI.Topology topo
+
+    Assert.assertNull(packingPlan);
+  }
+
```
extra newline

billonahill commented on heron/packing/tests/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPackingTest.java:
```diff
@@ -0,0 +1,570 @@
+  private double instanceCpuDefault;
+  private long instanceDiskDefault;
+
+
```
extra newline here and above

billonahill commented on heron/packing/tests/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPackingTest.java:
```diff
@@ -0,0 +1,570 @@
+    return count;
+  }
+
+  protected TopologyAPI.Topology getTopology(
```
do these 2 methods need to be protected?

avflor replied:
@billonahill This is the same as in current round robin packing test. Should I keep it as is or change it in both places?

billonahill commented on heron/packing/tests/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPackingTest.java:
```diff
@@ -0,0 +1,570 @@
+    PackingPlan packingPlan =
+        getResourceCompliantRRPackingPlan(topology);
+
+    Assert.assertNull(packingPlan);
```
This test should assert that a reasonable exception is thrown, instead of returning null

avflor replied:
Same as round robin. Should I keep it?

billonahill commented on heron/packing/tests/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPackingTest.java:
```diff
@@ -0,0 +1,570 @@
+        + (10 * (spoutParallelism + boltParallelism) * instanceDiskDefault) / 100;
+
+    Assert.assertEquals(packingPlanNoExplicitResourcesConfig.resource.disk, totalDisk);
+
```
extra newline

billonahill commented on heron/packing/tests/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPackingTest.java:
```diff
@@ -0,0 +1,570 @@
+    Assert.assertEquals(packingPlanNoExplicitResourcesConfig.containers.size(), 2);
+
+    long totalRam = (spoutParallelism + boltParallelism + 1) * instanceRamDefault
+        + (10 * (spoutParallelism + boltParallelism) * instanceRamDefault) / 100;
```
why +1 and *10 here? could 10 become a variable that implies what it is?

billonahill commented on heron/packing/tests/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPackingTest.java:
```diff
@@ -0,0 +1,570 @@
+
+    Assert.assertEquals(packingPlanExplicitResourcesConfig.containers.size(), numContainers);
+
+
```
extra newline

billonahill commented on heron/packing/tests/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPackingTest.java:
```diff
@@ -0,0 +1,570 @@
+  }
+
+  /**
+   * Test the scenario where container level resource config are set
```
How is testContainerRequestedResources2 different than testContainerRequestedResources? Could we name them as such or include a description in the javadocs?

billonahill commented on heron/packing/src/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPacking.java:
```diff
@@ -0,0 +1,419 @@
+  protected double maxContainerCpu;
+  protected long maxContainerDisk;
+  protected int numContainer;
+  protected int noAdjustments;
```
For consistency and clarity should change these to numContainers and numAdjustments. noAdjustments sounds like a boolean.

billonahill commented on heron/packing/src/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPacking.java:
```diff
@@ -0,0 +1,419 @@
+    for (String component : parallelismMap.keySet()) {
+      if (ramMap.containsKey(component)) {
+        if (!isValidInstance(ramMap.get(component), instanceCpuDefault, instanceDiskDefault)) {
+          return null;
```
what does returning null mean? If it's an error condition can we throw an exception?

billonahill commented on heron/packing/src/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPacking.java:
```diff
@@ -0,0 +1,419 @@
+    ArrayList<RamRequirement> ramRequirements = getRAMInstances();
+
+    if (ramRequirements == null) {
+      return null;
```
should clarify the semantics of what returning null means

avflor replied:
@billonahill Actually they are both severe cases and they are caused my misconfiguration of the topology parameters. I added a log message in both cases.

billonahill replied:
If this is due to invalid configs, throwing an exception seems preferred over returning null.

billonahill commented on heron/packing/src/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPacking.java:
```diff
@@ -0,0 +1,419 @@
+    int totalInstance = TopologyUtils.getTotalInstance(topology);
+
+    if (numContainer > totalInstance) {
+      throw new RuntimeException("More containers allocated than instance.");
```
let make this error message as descriptive as possible with things like the values that triggered the exception.

avflor replied:
@billonahill Same as round robin. Should I change in both places?

billonahill replied:
yes please.

billonahill commented on heron/packing/src/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPacking.java:
```diff
@@ -0,0 +1,419 @@
+    int index = 1;
+    int globalTaskIndex = 1;
+    Map<String, Integer> parallelismMap = TopologyUtils.getComponentParallelism(topology);
+    int j = 0;
```
should rename index and j to something more descriptive

avflor replied:
@billonahill Same as round robin

billonahill replied:
yes please.

billonahill commented on heron/packing/tests/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPackingTest.java:
```diff
@@ -0,0 +1,570 @@
+   * automatic adjustment has been enabled.
+   */
+  @Test
+  public void testInsufficientContainers3() throws Exception {
```
could you either name testInsufficientContainers[1|2|3] more descriptively or clarify in their javadocs what makes them different.

billonahill commented on heron/packing/tests/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPackingTest.java:
```diff
@@ -0,0 +1,570 @@
+
+    Assert.assertEquals(packingPlanExplicitResourcesConfig.containers.size(), numContainers);
+
+    Assert.assertEquals((long) (Math.round(4 * instanceCpuDefault
```
It seems this pattern of assertion for cpu/ram/disk is repeated in multiple tests. Could these be broken out into descriptive methods like assertCpu(...) that are reused?

avflor replied:
@billonahill As you may have noticed the equations are not the same across tests. It is hard to abstract them since they depend on the result of the packing algorithm so they are tailored for the specific cases used in these tests.

billonahill replied:
Got it. I thought they were similar except for the constants. The logic behind the math in the assertions is not entirely obvious so if there's something that could be done to help clarify the equation to the reader that would help.

avflor commented:
@billonahill @ajorgensen Thanks for the detailed feedback. Some parts of the code I copied from the current round robin algo and I'm not sure If I should change them or keep the same for consistency. I will mark individual changes so that you can clarify what is best to do.

avflor commented:
@billonahill @kramasamy @ajorgensen I think I addressed all the comments. I moved the Resource class out of the PackingPlan and this change affected multiple files. Please let me know if anything else is needed.

avflor commented:
@billonahill @kramasamy @ajorgensen BTW, is there any style formatter that applies the twitter guidelines like additional newlines? Checkstyle does not give me any errors on newlines and intellij does not fix the issue when reformatting the code. What do you usually use for that? Is there an intellij plugin for this?

kramasamy commented:
@avflor - we already enforce checkstyle when you compile. probably the empty line rule might not be there.

avflor commented:
@kramasamy Yeah, it's not there. Maybe we should add it. It will make the process much easier.

billonahill commented:
Yeah, it looks like some version of this one might be the one we need to checkstyles:
http://checkstyle.sourceforge.net/apidocs/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.html
https://github.com/twitter/heron/blob/master/tools/java/src/com/twitter/bazel/checkstyle/coding_style.xml

I'm not sure which one is needed for IntelliJ, but we'd set it here:
https://github.com/twitter/heron/blob/master/scripts/resources/idea/codeStyleSettings.xml

billonahill commented on heron/packing/src/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPacking.java:
```diff
@@ -0,0 +1,404 @@
+    ArrayList<RamRequirement> ramRequirements = getRAMInstances();
+
+    if (ramRequirements == null) {
+      throw new RuntimeException("The topology configuration does not have "
```
If this exception were thrown the reader would have little info about what went wrong. Instead could getRAMInstances throw a more detailed exception instead of returning null to this method, which then throws an exception? Best to throw as soon as an erroneous condition is found, which in this case is within getRAMInstances.

billonahill commented on heron/packing/tests/java/com/twitter/heron/packing/binpacking/FirstFitDecreasingPackingTest.java:
```diff
@@ -197,36 +229,36 @@ public void testContainerRequestedResources() throws Exception {
 
     Assert.assertEquals(Math.round(7 * instanceCpuDefault
-            + (10 * 7 * instanceCpuDefault) / 100 + instanceCpuDefault),
+            + (DEFAULT_CONTAINER_PADDING * 7 * instanceCpuDefault) / 100 + instanceCpuDefault),
```
Can we replace 7 with `totalInstances`? The more we clearly define the integers users, the easier it is to understand.

```int totalInstances = spoutParallelism + boltParallelism```

billonahill commented on heron/packing/tests/java/com/twitter/heron/packing/binpacking/FirstFitDecreasingPackingTest.java:
```diff
@@ -271,17 +303,17 @@ public void testCompleteRamMapRequested() throws Exception {
 
     Assert.assertEquals(4 * spoutRam + 3 * boltRam
-            + 10 * (4 * spoutRam + 3 * boltRam) / 100
+            + DEFAULT_CONTAINER_PADDING * (4 * spoutRam + 3 * boltRam) / 100
```
Instead of 4 and 3 could we use `spoutParallelism` and `boltParallelism` here (and elsewhere) respectively. Ideally we define ints with a member variable that states what it represents and then we only use the members, not the ints.

billonahill commented on heron/packing/tests/java/com/twitter/heron/packing/binpacking/FirstFitDecreasingPackingTest.java:
```diff
@@ -148,21 +178,23 @@ public void testDefaultContainerSizeWithPadding() throws Exception {
 
     double totalCpu = Math.round(spoutParallelism * instanceCpuDefault
-        + (50 * spoutParallelism * instanceCpuDefault) / 100)
+        + (padding * spoutParallelism * instanceCpuDefault) / 100)
```
This is a bit of a nit, but we divide by 100 for the padding right? In that case whenever we do that could we write this as `padding / 100` to make it more clear why the 100 exists?

avflor commented:
@billonahill I think I made all the changes. Regarding the constants in the tests:
A number 4 in an equation doesn't necessarily mean 4 spouts. It can reflect 2 spouts and 2 bolts in one container. So replacing the number with spoutParalellim (or BoltParallelsim) can be misleading. I made the change for the tests that this makes sense. I kept the constants for the tests that are more complex.

billonahill commented on heron/packing/tests/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPackingTest.java:
```diff
@@ -0,0 +1,540 @@
+    PackingPlan packingPlan =
+        getResourceCompliantRRPackingPlan(topology);
+
+    Assert.assertEquals(packingPlan.containers.size(), 2);
```
2 here is numContainers right? if so can we replace it?

billonahill commented on heron/packing/tests/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPackingTest.java:
```diff
@@ -0,0 +1,540 @@
+
+    Assert.assertEquals(packingPlan.resource.ram, totalRam);
+
+    double totalCpu = Math.round(4 * instanceCpuDefault
```
The 4s and 3s here appear to be spoutParallelism and boltParallelism. Is so can we replace them? Same for below.

avflor replied:
They are not spoutParallelism and BoltParallelism. In this test, the first container will contain 2 spouts and 2 bolts (4) and the second one (2 spouts and 1 bolt) so it cannot be replaced.

billonahill commented:
Thanks @avflor for being patient with my comments. :) The changes made have made things much easier to grok.

If the 4s and 3s don't align with bolt and spout counts, maybe a comment about what they are would do the trick. Without looking at the details of the algo it's hard to decipher why some of those numbers are being used in the test assertions.

avflor commented:
@billonahill I added some comments to clarify the numbers in the assertions. Hope it works now :)

kramasamy commented:
@ajorgensen - if you have any additional comments, let us know - so that we can merge it.

kramasamy commented:
@avflor - can you please resolve the conflicts?

avflor commented:
@kramasamy Fixed
